soroban-spec/src/gen/json.rs

```diff
@@ -0,0 +1,310 @@
+
+    fn try_from(spec: &ScSpecEntry) -> Result<Self, Self::Error> {
+        match spec {
+            ScSpecEntry::FunctionV0(f) => {
+                let name = f.name.to_string()?;
+                let inputs: Result<Vec<FunctionInput>, Error> =
+                    f.inputs.iter().map(FunctionInput::try_from).collect();
+                let outputs: Result<Vec<SpecType>, Error> =
+                    f.outputs.iter().map(SpecType::try_from).collect();
+                Ok(SpecType::Function {
+                    name: name,
+                    inputs: inputs?,
+                    outputs: outputs?,
+                })
+            }
+            ScSpecEntry::UdtStructV0(s) => {
+                let name = s.name.to_string()?;
+                let fields: Result<Vec<StructField>, Error> =
+                    s.fields.iter().map(StructField::try_from).collect();
+                Ok(SpecType::Struct {
+                    name: name,
+                    fields: fields?,
+                })
+            }
+            ScSpecEntry::UdtUnionV0(u) => {
+                let name = u.name.to_string()?;
+                let cases: Result<Vec<UnionCase>, Error> =
+                    u.cases.iter().map(UnionCase::try_from).collect();
+                Ok(SpecType::Union {
+                    name: name,
+                    cases: cases?,
+                })
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::SpecType;
+
+    const EXAMPLE_WASM: &[u8] =
+        include_bytes!("../../../target/wasm32-unknown-unknown/release/example_udt.wasm");
+
+    #[test]
+    fn example() {
+        let entries = crate::read::from_wasm(EXAMPLE_WASM).unwrap();
+        let mut json = "".to_string();
+        for entry in &entries {
+            let json_entry: SpecType = entry.try_into().unwrap();
+            json.push_str(&serde_json::to_string_pretty(&json_entry).unwrap());
+        }
+        assert_eq!(
+            json,
+            r#"{
+  "type": "union",
+  "name": "UdtEnum",
+  "cases": [
+    {
+      "name": "UdtA",
+      "value": null
+    },
+    {
+      "name": "UdtB",
+      "value": {
+        "type": "userDefined",
+        "name": "UdtStruct"
+      }
+    }
+  ]
+}{
+  "type": "struct",
+  "name": "UdtStruct",
+  "fields": [
+    {
+      "name": "a",
+      "value": {
+        "type": "i64"
+      }
+    },
+    {
+      "name": "b",
+      "value": {
+        "type": "i64"
+      }
+    },
+    {
+      "name": "c",
+      "value": {
+        "type": "vec",
+        "element": {
+          "type": "i64"
+        }
+      }
+    }
+  ]
+}{
+  "type": "function",
+  "name": "add",
+  "inputs": [
+    {
+      "name": "a",
+      "value": {
+        "type": "userDefined",
+        "name": "UdtEnum"
+      }
+    },
+    {
+      "name": "b",
+      "value": {
+        "type": "userDefined",
+        "name": "UdtEnum"
+      }
+    }
+  ],
+  "outputs": [
+    {
+      "type": "i64"
+    }
+  ]
+}"#
```

Comment:
@paulbellamy I think you were the stakeholder who asked for an easier to digest JSON format. Would you make any changes to this format? We could probably simplify it a little by removing the explicit 'userDefined' types and converting some objects inside arrays into objects. For example, the fields list could be a map/object of name to type. I don't feel strongly though.

Reply:
Yeah, here I think it's more important to be unambiguous without context. But I don't have a strong opinion either.